File: red/src/tools.ts
import localMain from "../resources/tools/ansible-local/main.yml" with { type: "text" };
import localInventory from "../resources/tools/ansible-local/inventory.ini" with { type: "text" };
import localCfg from "../resources/tools/ansible-local/ansible.cfg" with { type: "text" };
import * as sshConfig from './ssh-config.ts';
import {keyMode} from 'colors-compute-red';
import { plan_deployment, orchestrate, read_deployment, check_deployment_drift } from 'colors-compute-red';
import * as compute from './compute.ts';
// OpenTofu and Ansible stages for the fixed v1 topology, the port of
// io.github.getcolors.clickhouse.tools.

import { mkdirSync, writeFileSync } from "node:fs";
import { dirname, join } from "node:path";
import * as ansible from "red/ansible";
import { stageDir } from "red/cli";
import { toolEnv } from "red/providers";
import { PRESERVE_JINJA_DELIMITERS, contentSpec, scaffold, type Spec, type Template } from "red/scaffold";
import * as tofu from "red/tofu";
import { runtime } from "red/runtime";
import type { Opts } from "red/workflow";
import { StepError, failed } from "red/workflow";
import * as utils from "./utils.ts";
import * as validate from "./validate.ts";

import acceptancePy from "../resources/tools/acceptance/acceptance.py" with { type: "text" };
import ansibleCfg from "../resources/tools/ansible/ansible.cfg" with { type: "text" };
import ansibleCleanup from "../resources/tools/ansible/cleanup.yml" with { type: "text" };
import ansibleClickhouseConfig from "../resources/tools/ansible/clickhouse-config.xml" with { type: "text" };
import ansibleClickhouseUsers from "../resources/tools/ansible/clickhouse-users.xml" with { type: "text" };
import ansibleClickhouse from "../resources/tools/ansible/clickhouse.yml" with { type: "text" };
import ansibleCompose from "../resources/tools/ansible/docker-compose.yml" with { type: "text" };
import ansibleMain from "../resources/tools/ansible/main.yml" with { type: "text" };
import ansibleMetabase from "../resources/tools/ansible/metabase.yml" with { type: "text" };
import ansibleWireguard from "../resources/tools/ansible/wireguard.yml" with { type: "text" };
import dbtProject from "../resources/tools/dbt/dbt_project.yml" with { type: "text" };
import dbtEventsSummary from "../resources/tools/dbt/models/events_summary.sql" with { type: "text" };
import dbtSchema from "../resources/tools/dbt/models/schema.yml" with { type: "text" };
import dbtProfiles from "../resources/tools/dbt/profiles.yml" with { type: "text" };
import dbtPyproject from "../resources/tools/dbt/pyproject.toml" with { type: "text" };
import dbtEvents from "../resources/tools/dbt/seeds/events.csv" with { type: "text" };
import tofuDnsMainTf from "../resources/tools/tofu/dns/main.tf" with { type: "text" };

export const infrastructureTool = "clickhouse-infrastructure";
export const dnsTool = "clickhouse-dns";
export const ansibleTool = "clickhouse-ansible";
export const dbtTool = "clickhouse-dbt";
export const acceptanceTool = "clickhouse-acceptance";
export const tofuTools = [dnsTool];

export const templateOpts = PRESERVE_JINJA_DELIMITERS;

export function toolDir(opts: Opts, tool: string): string {
  return stageDir(opts, tool, { defaultProfile: "clickhouse" });
}

// The template tree this colour carries, keyed the way green names its
// classpath resources: "<path>/<file>" with dots as directories.
const templates: Record<string, string> = {
  "ansible-local/ansible.cfg":localCfg,"ansible-local/inventory.ini":localInventory,"ansible-local/main.yml":localMain,
  "acceptance/acceptance.py": acceptancePy,
  "ansible/ansible.cfg": ansibleCfg,
  "ansible/cleanup.yml": ansibleCleanup,
  // Bun's own types declare `*.xml` imports as Document; at runtime a
  // `with { type: "text" }` import is a string, so the cast restores the truth.
  "ansible/clickhouse-config.xml": ansibleClickhouseConfig as unknown as string,
  "ansible/clickhouse-users.xml": ansibleClickhouseUsers as unknown as string,
  "ansible/clickhouse.yml": ansibleClickhouse,
  "ansible/docker-compose.yml": ansibleCompose,
  "ansible/main.yml": ansibleMain,
  "ansible/metabase.yml": ansibleMetabase,
  "ansible/wireguard.yml": ansibleWireguard,
  "dbt/dbt_project.yml": dbtProject,
  "dbt/models/events_summary.sql": dbtEventsSummary,
  "dbt/models/schema.yml": dbtSchema,
  "dbt/profiles.yml": dbtProfiles,
  "dbt/pyproject.toml": dbtPyproject,
  "dbt/seeds/events.csv": dbtEvents,
  "tofu/dns/main.tf": tofuDnsMainTf,
};

export function template(path: string, file: string): Template {
  const name = `${path.replaceAll(".", "/")}/${file}`;
  const content = templates[name];
  if (content === undefined) throw new StepError(`template not found: ${name}`);
  return { name, content };
}

function spec(source: Template, target: string, data: Opts): Spec {
  return { template: source, target, data, opts: templateOpts };
}

const rawSpec = (target: string, content: string): Spec => contentSpec(target, content);

export function credentialEnv(opts: Opts, ...slots: string[]): Record<string, string> | undefined {
  return toolEnv(validate.providers, opts, [...slots, "provider-backend"]);
}

export async function tofuStep(opts: Opts, tool: string, specs: Spec[], slots: string[]): Promise<Opts> {
  const env = { ...credentialEnv(opts, ...slots), ...(opts["clickhouse/process-env"] ?? {}) };
  return tofu.tofuWithSpec(opts, specs, {
    dir: toolDir(opts, tool),
    env: Object.keys(env).length ? env : undefined,
  });
}

function sorted(value: any): any {
  if (Array.isArray(value)) return value.map(sorted);
  if (value && typeof value === 'object') return Object.fromEntries(Object.keys(value).sort().map(key=>[key,sorted(value[key])]));
  return value;
}
export async function infrastructureStep(opts: Opts): Promise<Opts> {
  try {
    const planning=opts['red/event']==='build'||opts['red/dry-run'];
    const result:any=planning?plan_deployment(opts,compute.TOPOLOGY,compute.requirements(opts)):await orchestrate(opts,compute.TOPOLOGY,compute.requirements(opts));
    if(planning) {
      const stages:any[]=[['shared',result.documents.shared],...Object.entries(result.documents.nodes).map(([node,docs])=>['nodes/'+node,docs])];
      for(const [stage,documents] of stages) for(const [name,document] of Object.entries(documents)) {
        const target=join(toolDir(opts,infrastructureTool),stage,name); mkdirSync(dirname(target),{recursive:true});
        writeFileSync(target,JSON.stringify(sorted(document),null,2)+'\n');
      }
    }
    if(!['ready','planned','destroyed'].includes(result.status)) return {...opts,'red/exit':1,'red/err':result.errors?.join('\n')||'compute lifecycle refused; inspect deployment state before retrying'};
    const output:Opts={...opts,'red/exit':0};
    if(result.cluster) Object.assign(output,{'colors-compute/cluster':result.cluster,'colors-compute/shared':result.shared});
    const path=result.key?.private_key_path;
    if(path) output['ssh-private-key-path']=planning?path.replace('$HOME/.ssh','/home/build-placeholder/.ssh'):path;
    return output;
  } catch {return {...opts,'red/exit':1,'red/err':'compute lifecycle refused; inspect deployment state before retrying'};}
}
export async function loadInfrastructureStep(opts: Opts): Promise<Opts> {
  if(opts['red/event']==='build'||opts['red/dry-run']) return infrastructureStep(opts);
  const result:any=await read_deployment(opts,undefined,undefined,compute.requirements(opts));
  if(result.status==='destroyed'&&opts['red/event']==='delete') return {...opts,'clickhouse/already-destroyed':true,'red/exit':0};
  if(result.status!=='present') return {...opts,'red/exit':1,'red/err':'compute state unavailable; legacy monolithic state requires explicit migration'};
  const output:Opts={...opts,'red/exit':0,'colors-compute/cluster':result.cluster,'colors-compute/shared':result.shared,'clickhouse/infrastructure-present?':true};
  if(result.key?.private_key_path) output['ssh-private-key-path']=result.key.private_key_path;
  return output;
}

export function dnsData(opts: Opts): Opts {
  return {
    ...opts,
    "metabase-host": utils.fqdn(opts, "metabase"),
    "clickhouse-host": utils.fqdn(opts, "clickhouse"),
  };
}

export async function dnsStep(opts: Opts): Promise<Opts> {
  const dir = toolDir(opts, dnsTool);
  return tofuStep(opts, dnsTool,
    [spec(template("tofu.dns", "main.tf"), `${dir}/main.tf`, dnsData(opts))],
    ["provider-dns"]);
}

export function allServers(opts: Opts): Record<string, Opts> {
  const nodes=Object.fromEntries(compute.resolved(opts).map(node=>[node.node_id,node]));
  return Object.fromEntries(utils.servers.map(app=>[app.id,{...app,...nodes[app['node-id']],'private-ip':nodes[app['node-id']].vpc_ip}]));
}

// Java's Double.toString, which is what Cheshire renders floats through and
// therefore what green's committed inventory bytes would carry. Integral
// numbers print as longs. JS's shortest-round-trip digits are the same digits
// Java chooses; only the layout differs.
function javaNumber(value: number): string {
  if (Number.isInteger(value)) return String(value);
  const negative = value < 0;
  const [mantissa, exponentPart] = Math.abs(value).toExponential().split("e");
  const exponent = Number(exponentPart);
  const digits = mantissa!.replace(".", "");
  let body: string;
  if (exponent >= -3 && exponent < 7) {
    if (exponent >= 0) {
      const intPart = digits.padEnd(exponent + 1, "0").slice(0, exponent + 1);
      const fracPart = digits.slice(exponent + 1);
      body = `${intPart}.${fracPart.length > 0 ? fracPart : "0"}`;
    } else {
      body = `0.${"0".repeat(-exponent - 1)}${digits}`;
    }
  } else {
    const rest = digits.slice(1);
    body = `${digits[0]}.${rest.length > 0 ? rest : "0"}E${exponent}`;
  }
  return negative ? `-${body}` : body;
}

// Cheshire's pretty printer, byte for byte: spaces around colons, arrays
// inline, nested objects newline-indented, floats in Java notation.
function pretty(value: unknown, indent = 0): string {
  if (Array.isArray(value)) {
    if (value.length === 0) return "[ ]";
    return `[ ${value.map((item) => pretty(item, indent)).join(", ")} ]`;
  }
  if (value !== null && typeof value === "object") {
    const entries = Object.entries(value);
    if (entries.length === 0) return "{ }";
    const pad = " ".repeat(indent + 2);
    return `{\n${entries
      .map(([key, nested]) => `${pad}${JSON.stringify(key)} : ${pretty(nested, indent + 2)}`)
      .join(",\n")}\n${" ".repeat(indent)}}`;
  }
  if (typeof value === "number") return javaNumber(value);
  return JSON.stringify(value ?? null);
}

export function inventory(opts: Opts): string {
  const servers = allServers(opts);
  const inventoryKey = opts["ssh-private-key-path"] ?? null;
  const hosts: Record<string, Opts> = {};
  for (const [id, s] of Object.entries(servers)) {
    hosts[utils.hostAlias(opts, id)] = {
      ansible_host: s.ip, ansible_user: s.user,
      private_ip: s["private-ip"], vpn_ip: s["vpn-ip"],
      server_role: s.role, server_ordinal: s.ordinal,
      ...(inventoryKey ? {ansible_ssh_private_key_file: inventoryKey} : {}),
    };
  }
  const selectKeys = (keys: string[]): Record<string, Opts> =>
    Object.fromEntries(keys.filter((key) => key in hosts).map((key) => [key, hosts[key]!]));
  return pretty({
    all: {
      children: {
        managed: { hosts },
        clickhouse: {
          hosts: selectKeys(utils.clickhouseServers().map((s) => utils.hostAlias(opts, s.id))),
        },
        metabase: { hosts: selectKeys([utils.hostAlias(opts, "metabase")]) },
        local: { hosts: { localhost: { ansible_connection: "local" } } },
      },
    },
  });
}

export function ansibleData(opts: Opts): Opts {
  return {
    ...opts,
    "metabase-host": utils.fqdn(opts, "metabase"),
    "clickhouse-host": utils.fqdn(opts, "clickhouse"),
    "local-wg-address": String(opts["wireguard-client-address"] ?? "").split("/")[0],
  };
}

export function ansibleSpecs(opts: Opts): Spec[] {
  const dir = toolDir(opts, ansibleTool);
  const data = ansibleData(opts);
  return [
    spec(template("ansible", "ansible.cfg"), `${dir}/ansible.cfg`, data),
    spec(template("ansible", "main.yml"), `${dir}/main.yml`, data),
    spec(template("ansible", "wireguard.yml"), `${dir}/wireguard.yml`, data),
    spec(template("ansible", "clickhouse.yml"), `${dir}/clickhouse.yml`, data),
    spec(template("ansible", "metabase.yml"), `${dir}/metabase.yml`, data),
    spec(template("ansible", "cleanup.yml"), `${dir}/cleanup.yml`, data),
    spec(template("ansible", "clickhouse-config.xml"), `${dir}/clickhouse-config.xml`, data),
    spec(template("ansible", "clickhouse-users.xml"), `${dir}/clickhouse-users.xml`, data),
    spec(template("ansible", "docker-compose.yml"), `${dir}/docker-compose.yml`, data),
    rawSpec(`${dir}/inventory.json`, inventory(opts)),
  ];
}

export function ansibleRenderStep(opts: Opts): Opts {
  return scaffold(opts, ansibleSpecs(opts));
}

export async function ansiblePlaybookStep(opts: Opts, playbook: string, recapKey: string): Promise<Opts> {
  if (opts["red/event"] === "build") return { ...opts, "red/exit": 0 };
  return ansible.ansibleStep(opts, {
    dir: toolDir(opts, ansibleTool),
    inventory: "inventory.json",
    playbooks: { create: playbook },
    hostKeyChecking: false,
    recapKey,
  });
}

export const wireguardStep = (opts: Opts) =>
  ansiblePlaybookStep(opts, "wireguard.yml", "clickhouse/wireguard-recap");
export const clickhouseConfigStep = (opts: Opts) =>
  ansiblePlaybookStep(opts, "clickhouse.yml", "clickhouse/clickhouse-recap");
export const metabaseConfigStep = (opts: Opts) =>
  ansiblePlaybookStep(opts, "metabase.yml", "clickhouse/metabase-recap");

export async function ansibleCleanupStep(opts: Opts): Promise<Opts> {
  return ansible.ansibleWithSpec(opts, {
    dir: toolDir(opts, ansibleTool),
    inventory: "inventory.json",
    playbooks: { delete: "cleanup.yml" },
    hostKeyChecking: false,
    recapKey: "clickhouse/cleanup-recap",
  }, ansibleSpecs(opts));
}

export async function dbtStep(opts: Opts): Promise<Opts> {
  const dir = toolDir(opts, dbtTool);
  const data = ansibleData(opts);
  const specs = [
    spec(template("dbt", "pyproject.toml"), `${dir}/pyproject.toml`, data),
    spec(template("dbt", "dbt_project.yml"), `${dir}/dbt_project.yml`, data),
    spec(template("dbt", "profiles.yml"), `${dir}/profiles.yml`, data),
    spec(template("dbt", "seeds/events.csv"), `${dir}/seeds/events.csv`, data),
    spec(template("dbt", "models/events_summary.sql"), `${dir}/models/events_summary.sql`, data),
    spec(template("dbt", "models/schema.yml"), `${dir}/models/schema.yml`, data),
  ];
  const rendered = scaffold(opts, specs);
  if (opts["red/event"] === "build" || opts["red/event"] === "delete") return rendered;
  const env = {
    DBT_PROFILES_DIR: dir,
    COLORS_DBT_PASSWORD: String(opts["clickhouse-dbt-password"] ?? ""),
  };
  const seed = await runtime.exec(["uv", "run", "dbt", "seed"], { cwd: dir, env });
  if (seed.exit !== 0) return { ...rendered, "red/exit": seed.exit, "red/err": seed.err };
  const run = await runtime.exec(["uv", "run", "dbt", "run", "--fail-fast"], { cwd: dir, env });
  if (run.exit !== 0) return { ...rendered, "red/exit": run.exit, "red/err": run.err };
  const test = await runtime.exec(["uv", "run", "dbt", "test"], { cwd: dir, env });
  if (test.exit !== 0) return { ...rendered, "red/exit": test.exit, "red/err": test.err };
  return rendered;
}

export async function acceptanceStep(opts: Opts): Promise<Opts> {
  const dir = toolDir(opts, acceptanceTool);
  const dbtDir = toolDir(opts, dbtTool);
  const data = ansibleData(opts);
  const script = `${dir}/acceptance.py`;
  const inventoryFile = join(toolDir(opts, ansibleTool), "inventory.json");
  const rendered = scaffold(opts, [spec(template("acceptance", "acceptance.py"), script, data)]);
  if (opts["red/event"] === "build" || opts["red/event"] === "delete") return rendered;
  const env = {
    COLORS_PAR_CLICKHOUSE_ADMIN_PASSWORD: String(opts["clickhouse-admin-password"] ?? ""),
    COLORS_PAR_CLICKHOUSE_METABASE_PASSWORD: String(opts["clickhouse-metabase-password"] ?? ""),
    COLORS_PAR_METABASE_ADMIN_EMAIL: String(opts["metabase-admin-email"] ?? ""),
    COLORS_PAR_METABASE_ADMIN_PASSWORD: String(opts["metabase-admin-password"] ?? ""),
  };
  const result = await runtime.exec(["uv", "run", "python", script, inventoryFile],
    { cwd: dbtDir, env });
  if (result.exit !== 0) return { ...rendered, "red/exit": result.exit, "red/err": result.err };
  return rendered;
}

export async function driftStep(opts: Opts): Promise<Opts> {
  if (opts["red/event"] !== "create") return { ...opts, "red/exit": 0 };
  const computeResult = await check_deployment_drift(opts, compute.TOPOLOGY, compute.requirements(opts));
  if(computeResult.status!=='clean') return {...opts,'red/exit':1,'red/err':computeResult.errors?.join('\n')||'compute drift check failed'};
  const env = credentialEnv(opts, "provider-dns");
  const results = await Promise.all(tofuTools.map(async (tool) =>
    [tool, await runtime.exec(
      ["tofu", `-chdir=${toolDir(opts, tool)}`, "plan", "-detailed-exitcode", "-input=false", "-no-color"],
      { env })] as const));
  const bad = results.find(([, result]) => result.exit !== 0);
  if (bad) {
    const [tool, result] = bad;
    return {
      ...opts, "red/exit": result.exit,
      "red/err": `OpenTofu drift remains in ${tool}\n${result.out}${result.err}`,
    };
  }
  return { ...opts, "red/exit": 0 };
}

export const ansibleLocalTool='clickhouse-ansible-local';
export function sshConfigHosts(opts:Opts){const nodes=compute.resolved(opts),entry=nodes.find(node=>node.node_id==='clickhouse-0');return [{...entry,name:opts.profile},...nodes.map(node=>({...node,name:opts.profile+'-'+node.node_id}))];}
export function ansibleLocalSpecs(opts:Opts):Spec[]{const data={...opts,'ssh-keygen':keyMode(opts).mode==='managed'},dir=toolDir(opts,ansibleLocalTool);return ['ansible.cfg','inventory.ini','main.yml'].map(name=>spec(template('ansible-local',name),dir+'/'+name,data));}
export async function ansibleLocalStep(opts:Opts):Promise<Opts>{
 if(opts['red/event']==='create'&&!opts['red/dry-run']){opts=sshConfig.preflight(opts);if(failed(opts))return opts;}
 return ansible.ansibleWithSpec(opts,{dir:toolDir(opts,ansibleLocalTool),inventory:'inventory.ini',playbooks:{create:'main.yml',delete:'main.yml'},extraVars:{host_alias:opts.profile,ssh_hosts:sshConfigHosts(opts),block_state:opts['red/event']==='delete'?'absent':'present'}},ansibleLocalSpecs(opts));
}

File: blue/src/package_clickhouse_blue/tools.py
"""OpenTofu and Ansible stages for the fixed v1 topology, the port of
io.github.getcolors.clickhouse.tools."""

from __future__ import annotations

import asyncio
import json
import math
import os
from decimal import Decimal
from pathlib import Path

from blue import tofu
from blue.ansible import ansible_step, ansible_with_spec
from blue.cli import stage_dir
from blue.providers import tool_env
from blue.runtime import runtime
from blue.scaffold import PRESERVE_JINJA_DELIMITERS, content_spec, scaffold
from blue.workflow import StepError, failed

from . import utils, validate, compute, ssh_config
from colors_compute.ssh import _mode as key_mode
from colors_compute.planning import plan_deployment
from colors_compute.orchestration import orchestrate
from colors_compute.inspection import read_deployment
from colors_compute.drift import check_deployment_drift

infrastructure_tool = 'clickhouse-infrastructure'
dns_tool = 'clickhouse-dns'
ansible_tool = 'clickhouse-ansible'
dbt_tool = 'clickhouse-dbt'
acceptance_tool = 'clickhouse-acceptance'
tofu_tools = [dns_tool]

ROOT = Path(__file__).parent / "resources"
template_opts = PRESERVE_JINJA_DELIMITERS


def tool_dir(opts: dict, tool: str) -> str:
    return stage_dir(opts, tool, default_profile="clickhouse")


def template(path: str, file: str) -> dict:
    name = f"tools/{path.replace('.', '/')}/{file}"
    return {"name": name, "content": (ROOT / name).read_text()}



def spec(source: dict, target: str, data: dict) -> dict:
    return {"template": source, "target": target, "data": data, "opts": template_opts}


def raw_spec(target: str, content: str) -> dict:
    return content_spec(target, content)


def credential_env(opts: dict, *slots: str) -> dict[str, str] | None:
    return tool_env(validate.providers, opts, [*slots, "provider-backend"])


async def tofu_step(opts: dict, tool: str, specs: list[dict], slots: list[str]) -> dict:
    env = {**(credential_env(opts, *slots) or {}),
           **(opts.get("clickhouse/process-env") or {})}
    return await tofu.tofu_with_spec(opts, specs, dir=tool_dir(opts, tool),
                                     env=env or None)


async def infrastructure_step(opts):
    try:
        planning = opts.get('blue/event') == 'build' or opts.get('blue/dry-run')
        result = plan_deployment(opts, compute.TOPOLOGY, compute.requirements(opts)) if planning else await orchestrate(opts, compute.TOPOLOGY, compute.requirements(opts))
        if planning:
            for stage, documents in [('shared', result['documents']['shared']), *[('nodes/' + node, documents) for node, documents in result['documents']['nodes'].items()]]:
                for name, document in documents.items():
                    target = Path(tool_dir(opts, infrastructure_tool)) / stage / name
                    target.parent.mkdir(parents=True, exist_ok=True)
                    target.write_text(json.dumps(document, indent=2, sort_keys=True) + '\n')
        if result['status'] not in ('ready', 'planned', 'destroyed'):
            return {**opts, 'blue/exit': 1, 'blue/err': '\n'.join(result.get('errors', [])) or 'compute lifecycle refused; inspect deployment state before retrying'}
        output = {**opts, 'blue/exit': 0}
        if result.get('cluster'):
            output.update({'colors-compute/cluster': result['cluster'], 'colors-compute/shared': result['shared']})
        path = result.get('key', {}).get('private_key_path')
        if path:
            output['ssh-private-key-path'] = path.replace('$HOME/.ssh', '/home/build-placeholder/.ssh') if planning else path
        return output
    except Exception:
        return {**opts, 'blue/exit': 1, 'blue/err': 'compute lifecycle refused; inspect deployment state before retrying'}


async def load_infrastructure_step(opts):
    if opts.get('blue/event') == 'build' or opts.get('blue/dry-run'):
        return await infrastructure_step(opts)
    result = await read_deployment(opts, None, None, compute.requirements(opts))
    if result['status'] == 'destroyed' and opts.get('blue/event') == 'delete':
        return {**opts, 'clickhouse/already-destroyed': True, 'blue/exit': 0}
    if result['status'] != 'present':
        return {**opts, 'blue/exit': 1, 'blue/err': 'compute state unavailable; legacy monolithic state requires explicit migration'}
    output = {**opts, 'blue/exit': 0, 'colors-compute/cluster': result['cluster'], 'colors-compute/shared': result['shared'], 'clickhouse/infrastructure-present?': True}
    path = result.get('key', {}).get('private_key_path')
    if path:
        output['ssh-private-key-path'] = path
    return output


def dns_data(opts: dict) -> dict:
    return {**opts,
            "metabase-host": utils.fqdn(opts, "metabase"),
            "clickhouse-host": utils.fqdn(opts, "clickhouse")}


async def dns_step(opts: dict) -> dict:
    dir = tool_dir(opts, dns_tool)
    return await tofu_step(opts, dns_tool,
                           [spec(template("tofu.dns", "main.tf"), f"{dir}/main.tf",
                                 dns_data(opts))],
                           ["provider-dns"])


def all_servers(opts):
    nodes = {node['node_id']: node for node in compute.resolved(opts)}
    return {app['id']: {**app, **nodes[app['node-id']], 'private-ip': nodes[app['node-id']]['vpc_ip']} for app in utils.servers}


def _java_double(x: float) -> str:
    """Java's Double.toString, which is what Green's cheshire JSON emits for
    floats: decimal between 1e-3 and 1e7, `d.dddE±e` scientific outside it.
    Python's own repr disagrees exactly where scientific notation starts
    (0.0001 -> "1.0E-4"), and the goldens carry the Java form."""
    if math.isnan(x):
        return "NaN"
    if math.isinf(x):
        return "Infinity" if x > 0 else "-Infinity"
    negative = math.copysign(1.0, x) < 0
    magnitude = abs(x)
    if magnitude == 0.0:
        return "-0.0" if negative else "0.0"
    _sign, digits, exponent = Decimal(repr(magnitude)).as_tuple()
    digit_str = "".join(map(str, digits)).rstrip("0") or "0"
    dec_exp = exponent + len(digits) - 1
    if -3 <= dec_exp < 7:
        if dec_exp >= 0:
            whole = digit_str[:dec_exp + 1].ljust(dec_exp + 1, "0")
            frac = digit_str[dec_exp + 1:] or "0"
        else:
            whole = "0"
            frac = "0" * (-dec_exp - 1) + digit_str
        rendered = f"{whole}.{frac}"
    else:
        mantissa = digit_str[0] + "." + (digit_str[1:] or "0")
        rendered = f"{mantissa}E{dec_exp}"
    return ("-" if negative else "") + rendered


def _pretty(value, indent=0):
    """Cheshire's pretty JSON, byte for byte — Green's artifact contract."""
    if isinstance(value, list):
        if not value:
            return "[ ]"
        return "[ " + ", ".join(_pretty(item, indent) for item in value) + " ]"
    if isinstance(value, dict):
        if not value:
            return "{ }"
        pad = " " * (indent + 2)
        body = ",\n".join(f"{pad}{json.dumps(str(k))} : {_pretty(v, indent + 2)}"
                          for k, v in value.items())
        return "{\n" + body + "\n" + " " * indent + "}"
    if isinstance(value, float) and not isinstance(value, bool):
        return _java_double(value)
    return json.dumps(value)


def inventory(opts: dict) -> str:
    servers = all_servers(opts)
    inventory_key = opts.get("ssh-private-key-path")
    hosts = {utils.host_alias(opts, id): {
        "ansible_host": s.get("ip"), "ansible_user": s.get("user"),
        "private_ip": s.get("private-ip"), "vpn_ip": s.get("vpn-ip"),
        "server_role": s.get("role"), "server_ordinal": s.get("ordinal"),
        **({"ansible_ssh_private_key_file": inventory_key} if inventory_key else {}),
    } for id, s in servers.items()}

    def select_keys(keys: list[str]) -> dict:
        return {key: hosts[key] for key in keys if key in hosts}

    return _pretty(
        {"all": {"children": {
            "managed": {"hosts": hosts},
            "clickhouse": {"hosts": select_keys(
                [utils.host_alias(opts, s["id"]) for s in utils.clickhouse_servers()])},
            "metabase": {"hosts": select_keys([utils.host_alias(opts, "metabase")])},
            "local": {"hosts": {"localhost": {"ansible_connection": "local"}}}}}})


def ansible_data(opts: dict) -> dict:
    address = opts.get("wireguard-client-address")
    return {**opts,
            "metabase-host": utils.fqdn(opts, "metabase"),
            "clickhouse-host": utils.fqdn(opts, "clickhouse"),
            "local-wg-address": ("" if address is None else str(address)).split("/")[0]}


def ansible_specs(opts: dict) -> list[dict]:
    dir = tool_dir(opts, ansible_tool)
    data = ansible_data(opts)
    return [spec(template("ansible", "ansible.cfg"), f"{dir}/ansible.cfg", data),
            spec(template("ansible", "main.yml"), f"{dir}/main.yml", data),
            spec(template("ansible", "wireguard.yml"), f"{dir}/wireguard.yml", data),
            spec(template("ansible", "clickhouse.yml"), f"{dir}/clickhouse.yml", data),
            spec(template("ansible", "metabase.yml"), f"{dir}/metabase.yml", data),
            spec(template("ansible", "cleanup.yml"), f"{dir}/cleanup.yml", data),
            spec(template("ansible", "clickhouse-config.xml"),
                 f"{dir}/clickhouse-config.xml", data),
            spec(template("ansible", "clickhouse-users.xml"),
                 f"{dir}/clickhouse-users.xml", data),
            spec(template("ansible", "docker-compose.yml"),
                 f"{dir}/docker-compose.yml", data),
            raw_spec(f"{dir}/inventory.json", inventory(opts))]


def ansible_render_step(opts: dict) -> dict:
    return scaffold(opts, ansible_specs(opts))


async def ansible_playbook_step(opts: dict, playbook: str, recap_key: str) -> dict:
    if opts.get("blue/event") == "build":
        return {**opts, "blue/exit": 0}
    return await ansible_step(opts, dir=tool_dir(opts, ansible_tool),
                              inventory="inventory.json",
                              playbooks={"create": playbook},
                              host_key_checking=False,
                              recap_key=recap_key)


async def wireguard_step(opts: dict) -> dict:
    return await ansible_playbook_step(opts, "wireguard.yml", "clickhouse/wireguard-recap")


async def clickhouse_config_step(opts: dict) -> dict:
    return await ansible_playbook_step(opts, "clickhouse.yml", "clickhouse/clickhouse-recap")


async def metabase_config_step(opts: dict) -> dict:
    return await ansible_playbook_step(opts, "metabase.yml", "clickhouse/metabase-recap")


async def ansible_cleanup_step(opts: dict) -> dict:
    return await ansible_with_spec(opts, ansible_specs(opts),
                                   dir=tool_dir(opts, ansible_tool),
                                   inventory="inventory.json",
                                   playbooks={"delete": "cleanup.yml"},
                                   host_key_checking=False,
                                   recap_key="clickhouse/cleanup-recap")


async def dbt_step(opts: dict) -> dict:
    dir = tool_dir(opts, dbt_tool)
    data = ansible_data(opts)
    specs = [spec(template("dbt", "pyproject.toml"), f"{dir}/pyproject.toml", data),
             spec(template("dbt", "dbt_project.yml"), f"{dir}/dbt_project.yml", data),
             spec(template("dbt", "profiles.yml"), f"{dir}/profiles.yml", data),
             spec(template("dbt", "seeds/events.csv"), f"{dir}/seeds/events.csv", data),
             spec(template("dbt", "models/events_summary.sql"),
                  f"{dir}/models/events_summary.sql", data),
             spec(template("dbt", "models/schema.yml"), f"{dir}/models/schema.yml", data)]
    rendered = scaffold(opts, specs)
    if opts.get("blue/event") in ("build", "delete"):
        return rendered
    password = opts.get("clickhouse-dbt-password")
    env = {"DBT_PROFILES_DIR": dir,
           "COLORS_DBT_PASSWORD": "" if password is None else str(password)}
    for args in [["uv", "run", "dbt", "seed"],
                 ["uv", "run", "dbt", "run", "--fail-fast"],
                 ["uv", "run", "dbt", "test"]]:
        result = await runtime.exec(args, cwd=dir, env=env)
        if result.exit != 0:
            return {**rendered, "blue/exit": result.exit, "blue/err": result.err}
    return rendered


async def acceptance_step(opts: dict) -> dict:
    dir = tool_dir(opts, acceptance_tool)
    dbt_dir = tool_dir(opts, dbt_tool)
    data = ansible_data(opts)
    script = f"{dir}/acceptance.py"
    inventory_file = os.path.join(tool_dir(opts, ansible_tool), "inventory.json")
    rendered = scaffold(opts, [spec(template("acceptance", "acceptance.py"), script, data)])
    if opts.get("blue/event") in ("build", "delete"):
        return rendered

    def par(key: str) -> str:
        value = opts.get(key)
        return "" if value is None else str(value)

    env = {"COLORS_PAR_CLICKHOUSE_ADMIN_PASSWORD": par("clickhouse-admin-password"),
           "COLORS_PAR_CLICKHOUSE_METABASE_PASSWORD": par("clickhouse-metabase-password"),
           "COLORS_PAR_METABASE_ADMIN_EMAIL": par("metabase-admin-email"),
           "COLORS_PAR_METABASE_ADMIN_PASSWORD": par("metabase-admin-password")}
    result = await runtime.exec(["uv", "run", "python", script, inventory_file],
                                cwd=dbt_dir, env=env)
    if result.exit != 0:
        return {**rendered, "blue/exit": result.exit, "blue/err": result.err}
    return rendered


async def drift_step(opts: dict) -> dict:
    if opts.get("blue/event") != "create":
        return {**opts, "blue/exit": 0}
    compute_drift = await check_deployment_drift(opts, compute.TOPOLOGY, compute.requirements(opts))
    if compute_drift.get("status") != "clean":
        return {**opts, "blue/exit": 1, "blue/err": "compute drift remains or could not be checked"}
    env = credential_env(opts, "provider-dns")

    async def plan(tool: str):
        return (tool, await runtime.exec(
            ["tofu", f"-chdir={tool_dir(opts, tool)}",
             "plan", "-detailed-exitcode", "-input=false", "-no-color"],
            env=env))

    results = await asyncio.gather(*(plan(tool) for tool in tofu_tools))
    bad = next(((tool, result) for tool, result in results if result.exit != 0), None)
    if bad:
        tool, result = bad
        return {**opts, "blue/exit": result.exit,
                "blue/err": f"OpenTofu drift remains in {tool}\n{result.out}{result.err}"}
    return {**opts, "blue/exit": 0}


ansible_local_tool = 'clickhouse-ansible-local'

def ssh_config_hosts(opts):
    nodes = compute.resolved(opts)
    entry = next(node for node in nodes if node['node_id'] == 'clickhouse-0')
    return [{**entry,'name':opts['profile']}, *[{**node,'name':opts['profile']+'-'+node['node_id']} for node in nodes]]

def ansible_local_specs(opts):
    data = {**opts, 'ssh-keygen':key_mode(opts)['mode']=='managed'}
    directory = tool_dir(opts,ansible_local_tool)
    return [spec(template('ansible-local',name),directory+'/'+name,data) for name in ['ansible.cfg','inventory.ini','main.yml']]

async def ansible_local_step(opts):
    if opts.get('blue/event')=='create' and not opts.get('blue/dry-run'):
        opts=ssh_config.preflight(opts)
        if failed(opts): return opts
    return await ansible_with_spec(opts,ansible_local_specs(opts),dir=tool_dir(opts,ansible_local_tool),inventory='inventory.ini',playbooks={'create':'main.yml','delete':'main.yml'},extra_vars={'host_alias':opts['profile'],'ssh_hosts':ssh_config_hosts(opts),'block_state':'absent' if opts.get('blue/event')=='delete' else 'present'})
